### ISSUE TITLE
bugprone-branch-clone false positive on template instantiations

### ISSUE THREAD

**Issue by njames93** [llvm/llvm-project#62693]
The bugprone-branch-clone check currently emits warnings when template instantiations can result in a branch clone

Take this example here
```cpp
class Object {
   public:
    template <typename T>
    bool ConvertableTo() const;

    template <typename T>
    void Handle();
};

template <typename T>
void update(Object &a) {
    if (a.ConvertableTo<char *>()) {
        a.Handle<char *>();
    } else {
        a.Handle<T>();
    }
}

template <typename T>
void update2(Object &a) {
    if (a.ConvertableTo<char *>()) {
        a.Handle<char *>();
    } else {
        a.Handle<T>();
    }
}

void foo(Object &a) {
    update<int>(a);
    update2<char *>(a);
}
```
In this code, there is no warning emitted from `update`, which has exactly the same code as `update2`. However due to the instantiation of `update2` with type `char *`, A warning is emitted on the line containing the `if` statement in `update2`.

https://godbolt.org/z/Pefv9E6Mv

**Comment by phyBrackets** [comment]
Why would it warns for `update` branches? the `update` function is instantiated with type int, so there is no issue with branching on the if statement, they doesn't seems to be the identical branching case? Or am i missing something?

**Comment by njames93** [comment]
It shouldn't warn for either, it's warning for `update2` because the instantiated code is identical, however the template definition code is not identical branches.

**Comment by PiotrZSL** [comment]
Agree, implicit code shouldn't be checked. Most probably check need to be corrected by changing:
`ifStmt(unless(allOf(isConstexpr(), isInTemplateInstantiation())),`
into
`ifStmt(unless(isInTemplateInstantiation())),`

**Comment by PiotrZSL** [comment]
https://reviews.llvm.org/D151133
